text/0000-dual-light-shadow.md

```diff
@@ -0,0 +1,297 @@
+---
+title: Dual light/shadow components
+status: DRAFTED
+created_at: 2022-02-08
+updated_at: 2022-02-08
+champion: Nolan Lawson (nolanlawson)
+pr: https://github.com/salesforce/lwc-rfcs/pull/77
+---
+
+# Dual light/shadow components
+
+## Summary
+
+[Light DOM components](https://rfcs.lwc.dev/rfcs/lwc/0115-light-dom) are a useful feature in certain contexts.
+However, LWC currently requires a component to be either light or shadow – it cannot be both at once.
+
+This RFC proposes a "dual" mode where the same component can run in either light DOM mode or shadow DOM mode.
+
+## Basic example
+
+A component can declare its `lwc:render-mode` to be `"dual"`:
+
+```html
+<!-- component.html -->
+<template lwc:render-mode="dual">
+    <h1>Hello world</h1>
+</template>

```

Comment:
@AllanOricil Thanks for the feedback! I appreciate it a lot. 🙂 

> couldn't the render-mode directive be simplified to a boolean

I believe the reason we went with a string enum originally was because we weren't sure if someday we would want to support more exotic cases, like closed shadow mode or ["open-stylable" shadow mode](https://github.com/WICG/webcomponents/issues/909). Although maybe that would argue that we should use a different name for `"dual"` – perhaps `"any"` or something.

> What if dual is the default render mode for every component?

I can see the case for "dual by default," but the main issue I see is that components that were not designed with light DOM in mind (or with shadow DOM in mind, for light DOM components) may be forced into a mode that breaks them. Whereas with my proposal, light/shadow components only render in that one mode unless they explicitly opt-in.

>  I believe that using a variable called `useLightDom` inside the `render()`

This would be very convenient from the component authoring perspective, but unfortunately it would greatly complicate how the engine is implemented for the component to switch between light and shadow mode dynamically during its lifetime. For instance, if light-DOM styles were added to the `<head>`, now those have to be converted into shadow DOM styles in the shadow root (or vice versa). And while it's possible to attach a new shadow root to an existing light DOM component,  it's not possible to remove a shadow root (the browser does not provide a "remove" analog to [`attachShadow`](https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow)).

> 3rd party companies would have document which components render in light mode no matter the context, because developers wouldn't be able to determine this information by just reading the parent component's template.

I don't think they would need to document it either way, because we should be able to expose the static `renderMode` in the [metadata](https://www.npmjs.com/package/@lwc/metadata). So it would be available to e.g. a drag-and-drop builder UI.

Reply:
> What if dual is the default render mode for every component?

That is non-backward compatible for you. Additionally, we could say that you can, as a consumer of the component, always only select the mode that you know the consumer component works... saying that it is a responsibility of the consumer to always specify has two big drawbacks that I can't really sign off on:

1. if the current version of the component does work in both modes (just by pure luck), and the consumer select light, then the owner of the component modifies that, and inadvertently breaks light, he never really sign on for that, now we have breaking changes.
2. now the consumer has to learn that there is to mode, and which components are broken and needs to be forced to a particular mode of operation, that seems very wrong to me.

Reply:
1- I think this could be mitigated if 3rd party developers release a MAJOR version for their package. Salesforce could instruct 3rd party developers to do this when forcing a render mode on their components, if they were previously "dual". Consumers would have to the read release notes to understand clearly why a MAJOR bump happened, and what could be the implications before upgrading their package.

Reply:
I understand it now 👍